### PR TITLE
Automaticly generate textures when binding new texture

### PR DESCRIPTION
**Pull request by mittorn**
This is useful if application calls glBindTexture and sets some glTexParameters before uploading it
https://github.com/kripken/emscripten/issues/4984

### CODE REVIEW

File: src/library_gl.js [hunk 1]
```diff
@@ -1257,7 +1257,12 @@ var LibraryGL = {
 #if GL_ASSERTIONS
     GL.validateGLObjectID(GL.textures, texture, 'glBindTexture', 'texture');
 #endif
-    GLctx.bindTexture(target, texture ? GL.textures[texture] : null);
+    if( texture >= 0 )
```
Comment (kripken) [inline src/library_gl.js]: indentation and formatting should match the rest of the code, that is, `if (texture >= 0) {` on the first line etc.

**Comment by kripken** [comment]
I don't understand what this fixes, do you have a testcase showing the bug?

**Comment by mittorn** [comment]
Sorry for identation error, i didn't notice it.
I'm getting 502 error when loading texture without calling glGenTextures or glCreateTexture
Texture loading procedure is similar to this
```
glBindTexture(GL_TEXTURE_2D,0);
glTexParametri(...);
glTexImage2D(...);
```
This method used in many gl1 applications, but in WebGL glCreateTexture required to get valid texture id.

**Comment by kripken** [comment]
Reading the docs, looks like this is [true in GL1](https://www.talisman.org/opengl-1.1/Reference/glBindTexture.html) and [false in GL4](https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glBindTexture.xhtml)? Then this should be behind the `LEGACY_GL_EMULATION` flag, as this is a case of emulating older GL behaviors.

Please also make the PR to the incoming branch, and add a test (see other tests using `LEGACY_GL_EMULATION` in tests/test_browser.py).

**Comment by mittorn** [comment]
"You may use" in gl es 2.0
https://www.khronos.org/registry/OpenGL-Refpages/es2.0/
Gl es 2.0 projects do not need any emulation if vbo used.
Maybe implement it as a separate option?

File: src/library_gl.js [hunk 2]
```diff
@@ -1257,7 +1257,12 @@ var LibraryGL = {
...
+    {
+    if( GL.textures[texture] == undefined )
```
Comment (juj) [inline src/library_gl.js]: Instead of checking against `undefined`, better to do `if (!GL.textures[texture]) {`

**Comment by juj** [comment]
Yeah, the implicit creation of GL resources on first bind is a horrible horrible feature from Desktop OpenGL, which was eventually removed in OpenGL 3.1 specification (https://en.wikipedia.org/wiki/OpenGL#OpenGL_3.1, https://www.khronos.org/registry/OpenGL/specs/gl/glspec31.pdf), the earlier Desktop OpenGL 3.0 spec still allowed creating resources on first bind. Unfortunately this deprecation never made its way to OpenGL ES land and the latest OpenGL ES 3.2 spec still has this implicit create on bind feature.

I don't want to support this by default, nor in `FULL_ES2=1` mode, because way more often than not it is a bug on programmer's part rather than intended use, so enabling this by default would potentially hide real bugs, but I'm very happy to merge and maintain support for this as long as
  a) it is systematically done for all resource binds that the GLES2&3 specs say it should be supported for (which other glBinds should create resource if they did not exist?), and
  b) we put it behind a special flag. Does not need to be `LEGACY_GL_EMULATION`, but perhaps `GL_CREATE_RESOURCE_ON_BIND=0/1` (default = 0),
  c) this PR has a bug that if one creates multiple resources implicitly like this, then later `glGen*` functions will currently not see that those names have already been used, and https://github.com/kripken/emscripten/blob/master/src/library_gl.js#L88 will generate duplicate names that will collide with this. Need to update that code to check that the automatically generated IDs will not be in use, and
  d) please rebase this PR against `incoming` branch and add yourself to `AUTHORS` if not already?

File: src/library_gl.js (hunk 2 above)
Comment (mittorn) [inline src/library_gl.js]: but 0 is valid gl texture name to

**Comment by kripken** [comment]
> we put it behind a special flag. Does not need to be LEGACY_GL_EMULATION, but perhaps GL_CREATE_RESOURCE_ON_BIND=0/1 (default = 0),

Sounds good to me too. How about `LEGACY_GL_CREATE_RESOURCE_ON_BIND`? (i.e., add `LEGACY_` as a prefix?

**Comment by juj** [comment]
> Sounds good to me too. How about LEGACY_GL_CREATE_RESOURCE_ON_BIND? (i.e., add LEGACY_ as a prefix?)

Yeah, that would be fine as well.

**Pull request closed without merging**